### PR TITLE
Service port compatibility check accounts for TCP/UDP differentiation

### LINKED ISSUE #34677: LB-IPAM: Account for protocol differentiation in IP sharing checks
LB-IPAM has a feature which allows services to share an IP under a number of conditions. One of these is that ports are not allowed to overlap. So far, it didn't matter for which protocol these ports were since the Cilium datapath was unable to differentiate between TCP and UDP. Now that https://github.com/cilium/cilium/pull/33434 has merged, we should adjust this check so a service with port UDP/80 and and a service with port TCP/80 can share an IP.

The logic in question lives here: https://github.com/cilium/cilium/blob/6b211cc4c882cb32bed933ffd5287a7f830ee714/operator/pkg/lbipam/service_store.go#L107

### PR DESCRIPTION
**Pull request by ldlb9527**
fixes: #34677 

```release-note
Make LB-IPAM allow IP sharing between services with the same ports but different protocols
```

### CODE REVIEW

**Comment by ysksuzuki** [comment]
@dylandreimerink Would you mind taking look at this PR? You have the most context on #34677

File: operator/pkg/lbipam/service_store.go [hunk 1]
```diff
@@ -101,12 +102,14 @@ func (sv *ServiceView) isCompatible(osv *ServiceView) (bool, string) {
...
-	// 	     protocols, so we don't either for this purpose. https://github.com/cilium/cilium/issues/9207
+	// Compatible services don't have any overlapping ports with the same protocol.
+	// NOTE: The Cilium datapath can differentiate between protocols,thanks to the merge
+	//       of PR https://github.com/cilium/cilium/pull/33434.
+	diff := option.Config.LoadBalancerProtocolDifferentiation
 	for _, port1 := range sv.Ports {
 		for _, port2 := range osv.Ports {
-			if port1.Port == port2.Port {
+			if port1.Port == port2.Port &&
+				(!diff || (diff && port1.Protocol == port2.Protocol && (port1.Protocol == slim_core_v1.ProtocolTCP || port1.Protocol == slim_core_v1.ProtocolUDP))) {
```
Comment (dylandreimerink) [inline operator/pkg/lbipam/service_store.go]: I think this is correct, but its very hard to verify, had to draw out the truth table of 16 combinations. 

I also double checked https://github.com/cilium/cilium/pull/33434, while it is explicit about UDP and TCP, it should also work for other L4 protocols such as [SCTP](https://docs.cilium.io/en/latest/configuration/sctp/) so no need to special case the protocols, making this easier.

I would suggest the following, it isn't optimized formal logic, and will not make any mathematician happy, but its easy to understand.

```suggestion
                        portsEqual := port1.Port == port2.Port
                        protocolEqual := port1.Protocol == port2.Protocol
			if diff && portsEqual && protocolEqual { 
				return false, "same port and protocol"
			}
			if !diff && portsEqual {
```

**Review by dylandreimerink — CHANGES_REQUESTED** [review]
Thank you! This is great, just want to simplify a little

**Comment by dylandreimerink** [comment]
I have two requests that I forgot to add to the initial review. The first being, could you please add a test case. We already have one that asserts the non-overlapping ports, you should be able to copy it and modify it, set the service differentiation setting in the new case and make the two services use UDP and TCP on the same port, assert they get the same IP assigned. https://github.com/cilium/cilium/blob/1cb9b66182d8c0974e83161323ee6c76afdf4bb6/operator/pkg/lbipam/lbipam_test.go#L241

My second request is to slightly change the way we plumb in the service differentiation setting. The current way uses the global variable, which may cause issues during testing since changing it persists across test cases. What I suggest instead is to use the config object that is already provided. https://github.com/cilium/cilium/blob/1cb9b66182d8c0974e83161323ee6c76afdf4bb6/operator/pkg/lbipam/cell.go#L59

You can add a new boolean [here](https://github.com/cilium/cilium/blob/1cb9b66182d8c0974e83161323ee6c76afdf4bb6/operator/pkg/lbipam/lbipam.go#L74) for the setting and set it [here](https://github.com/cilium/cilium/blob/1cb9b66182d8c0974e83161323ee6c76afdf4bb6/operator/pkg/lbipam/cell.go#L87)

Finally, since the logic is in ServiceStore, add a new boolean to the service store, and its constructor [here](https://github.com/cilium/cilium/blob/1cb9b66182d8c0974e83161323ee6c76afdf4bb6/operator/pkg/lbipam/service_store.go#L19-L31) then plum the setting in [here](https://github.com/cilium/cilium/blob/1cb9b66182d8c0974e83161323ee6c76afdf4bb6/operator/pkg/lbipam/lbipam.go#L118)

I hope that made sense :sweat_smile:. Its a few additional lines of code, but it makes this piece of code easier to test without the global variable usage, so worth it

File: operator/pkg/lbipam/service_store.go (hunk 1 above)
Comment (ldlb9527) [reply operator/pkg/lbipam/service_store.go]: Thank you for your suggestion, I agree that code readability is important.

**Comment by ldlb9527** [comment]
hi @dylandreimerink, I have improved this PR, but I didn’t use the [DaemonConfig](https://github.com/cilium/cilium/blob/1cb9b66182d8c0974e83161323ee6c76afdf4bb6/operator/pkg/lbipam/cell.go#L59) you mentioned. I’m not sure if this is correct.

**Review by dylandreimerink — APPROVED** [review]
This is what I was after, thank you!

**Review by tommyp1ckles — APPROVED** [review]
looks good, thank you

**Review by ldelossa — APPROVED** [review]
LGTM. Thanks!

**Comment by julianwiedmann** [comment]
Let's see if a rebase helps to resolve the one CI fail in e2e-upgrade.

**Pull request merged**